File: rogalik.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <conio.h>  // для _getch() на Windows
// #include <unistd.h>  // для getch() на Linux

#ifdef _WIN32
#include <windows.h>
#endif

using namespace std;

class Roguelike {
private:
    const int WIDTH = 40;
    const int HEIGHT = 20;
    const int FOV_RADIUS = 5;  // радиус обзора фонарика
    
    vector<vector<char>> map;
    int playerX, playerY;
    bool gameRunning;
    
    enum Tile {
        EMPTY = ' ',
        WALL = '#',
        FLOOR = '.',
        PLAYER = '@',
        ENEMY = 'E',
        TREASURE = 'T'
    };
    
public:
    Roguelike() {
        srand(time(0));
        map.resize(HEIGHT, vector<char>(WIDTH, EMPTY));
        playerX = 1;
        playerY = 1;
        gameRunning = true;
        generateMap();
    }
    
    void generateMap() {
        // Создаем случайные комнаты и коридоры
        for (int y = 0; y < HEIGHT; y++) {
            for (int x = 0; x < WIDTH; x++) {
                if (x == 0  y == 0  x == WIDTH-1 || y == HEIGHT-1) {
                    map[y][x] = WALL;
                } else if (rand() % 100 < 30) {  // 30% chance для стен
                    map[y][x] = WALL;
                } else {
                    map[y][x] = FLOOR;
                }
            }
        }
        
        // Добавляем игрока
        map[playerY][playerX] = PLAYER;
        
        // Добавляем врагов
        for (int i = 0; i < 5; i++) {
            int x, y;
            do {
                x = rand() % (WIDTH-2) + 1;
                y = rand() % (HEIGHT-2) + 1;
            } while (map[y][x] != FLOOR);
            map[y][x] = ENEMY;
        }
        
        // Добавляем сокровища
        for (int i = 0; i < 3; i++) {
            int x, y;
            do {
                x = rand() % (WIDTH-2) + 1;
                y = rand() % (HEIGHT-2) + 1;
            } while (map[y][x] != FLOOR);
            map[y][x] = TREASURE;
        }
        
        // Гарантируем проходимость вокруг игрока
        for (int dy = -1; dy <= 1; dy++) {
            for (int dx = -1; dx <= 1; dx++) {
                if (playerY+dy >= 0 && playerY+dy < HEIGHT && 
                    playerX+dx >= 0 && playerX+dx < WIDTH) {
                    map[playerY+dy][playerX+dx] = FLOOR;
                }
            }
        }
        map[playerY][playerX] = PLAYER;
    }
    
    void clearScreen() {
        #ifdef _WIN32
        system("cls");
        #else
        system("clear");
        #endif
    }
    
    bool isInFOV(int x, int y) {
        // Простая проверка расстояния для фонарика
        int dx = x - playerX;
        int dy = y - playerY;
        return (dx*dx + dy*dy) <= FOV_RADIUS * FOV_RADIUS;
    }
    
    void render() {
        clearScreen();
        
        cout << "=== КОНСОЛЬНЫЙ РОГАЛИК С ФОНАРИКОМ ===" << endl;
        cout << "Управление: WASD - движение, Q - выход" << endl;
        cout << "Фонарик освещает область вокруг игрока" << endl;
        cout << "Символы: @ - вы, # - стены, E - враги, T - сокровища" << endl;
        cout << string(WIDTH + 2, '-') << endl;
        
        for (int y = 0; y < HEIGHT; y++) {
            cout << '|';
            for (int x = 0; x < WIDTH; x++) {
                if (isInFOV(x, y) || map[y][x] == PLAYER) {
                    cout << map[y][x];
                } else {
                    cout << ' ';  // неосвещенная область
                }
            }
            cout << '|' << endl;
        }
        
        cout << string(WIDTH + 2, '-') << endl;
    }
    
    bool isValidMove(int x, int y) {
        return x >= 0 && x < WIDTH && y >= 0 && y < HEIGHT && map[y][x] != WALL;
    }
    
    void movePlayer(int dx, int dy) {
        int newX = playerX + dx;
        int newY = playerY + dy;if (isValidMove(newX, newY)) {
            // Проверяем, что на новой клетке
            if (map[newY][newX] == ENEMY) {
                cout << "Вы встретили врага! Игра окончена." << endl;
                gameRunning = false;
                return;
            } else if (map[newY][newX] == TREASURE) {
                cout << "Вы нашли сокровище! Поздравляем!" << endl;
                gameRunning = false;
                return;
            }
            
            // Перемещаем игрока
            map[playerY][playerX] = FLOOR;
            playerX = newX;
            playerY = newY;
            map[playerY][playerX] = PLAYER;
        }
    }
    
    void processInput() {
        char input = _getch();  // на Windows
        // char input = getch();  // на Linux
        
        switch (tolower(input)) {
            case 'w': movePlayer(0, -1); break;
            case 's': movePlayer(0, 1); break;
            case 'a': movePlayer(-1, 0); break;
            case 'd': movePlayer(1, 0); break;
            case 'q': gameRunning = false; break;
        }
    }
    
    void run() {
        while (gameRunning) {
            render();
            processInput();
            
            // Небольшая задержка для плавности
            #ifdef _WIN32
            Sleep(50);
            #else
            usleep(50000);
            #endif
        }
        
        cout << "Спасибо за игру!" << endl;
    }
};

int main() {
    // Настройка кодировки для Windows
    #ifdef _WIN32
    SetConsoleOutputCP(65001);
    #endif
    
    Roguelike game;
    game.run();
    return 0;
}

File: rogalik.py
import os
import random
import math
import time

class Roguelike:
    def __init__(self):
        self.WIDTH = 40
        self.HEIGHT = 20
        self.FOV_RADIUS = 5
        
        self.EMPTY = ' '
        self.WALL = '#'
        self.FLOOR = '.'
        self.PLAYER = '@'
        
        self.map = []
        self.visible = []
        self.explored = []
        self.playerX = 1
        self.playerY = 1
        self.gameRunning = True
        
        self.initialize_arrays()
        self.generate_map()
        self.update_fov()
    
    def initialize_arrays(self):
        self.map = [[self.EMPTY for _ in range(self.WIDTH)] for _ in range(self.HEIGHT)]
        self.visible = [[False for _ in range(self.WIDTH)] for _ in range(self.HEIGHT)]
        self.explored = [[False for _ in range(self.WIDTH)] for _ in range(self.HEIGHT)]
    
    def cast_ray(self, x1, y1, x2, y2):
        dx = abs(x2 - x1)
        dy = abs(y2 - y1)
        sx = 1 if x1 < x2 else -1
        sy = 1 if y1 < y2 else -1
        err = dx - dy
        
        x, y = x1, y1
        
        while True:
            if 0 <= x < self.WIDTH and 0 <= y < self.HEIGHT:
                self.visible[y][x] = True
                self.explored[y][x] = True
                
                if self.map[y][x] == self.WALL:
                    break
            
            if x == x2 and y == y2:
                break
            
            e2 = 2 * err
            if e2 > -dy:
                err -= dy
                x += sx
            if e2 < dx:
                err += dx
                y += sy
    
    def update_fov(self):
        for y in range(self.HEIGHT):
            for x in range(self.WIDTH):
                self.visible[y][x] = False
        
        self.visible[self.playerY][self.playerX] = True
        self.explored[self.playerY][self.playerX] = True
        
        for angle in range(0, 360, 5):
            rad = angle * math.pi / 180.0
            endX = self.playerX + int(self.FOV_RADIUS * math.cos(rad))
            endY = self.playerY + int(self.FOV_RADIUS * math.sin(rad))
            self.cast_ray(self.playerX, self.playerY, endX, endY)
    
    def generate_map(self):
        for y in range(self.HEIGHT):
            for x in range(self.WIDTH):
                if x == 0 or y == 0 or x == self.WIDTH-1 or y == self.HEIGHT-1:
                    self.map[y][x] = self.WALL
                elif random.randint(0, 99) < 30:
                    self.map[y][x] = self.WALL
                else:
                    self.map[y][x] = self.FLOOR
        
        self.map[self.playerY][self.playerX] = self.PLAYER
        self.clear_area_around_player()
    
    def clear_area_around_player(self):
        for dy in range(-1, 2):
            for dx in range(-1, 2):
                newY = self.playerY + dy
                newX = self.playerX + dx
                if 0 <= newY < self.HEIGHT and 0 <= newX < self.WIDTH:
                    self.map[newY][newX] = self.FLOOR
        self.map[self.playerY][self.playerX] = self.PLAYER
    
    def clear_screen(self):
        os.system('cls' if os.name == 'nt' else 'clear')
    
    def render(self):
        self.clear_screen()
        
        print("Управление: WASD - движение, Q - выход")
        print("Символы: @ - вы, # - стены")
        
        for y in range(self.HEIGHT):
            for x in range(self.WIDTH):
                if self.visible[y][x]:
                    print(self.map[y][x], end='')
                elif self.explored[y][x]:
                    if self.map[y][x] == self.WALL:
                        print('#', end='')
                    else:
                        print('.', end='')
                else:
                    print(' ', end='')
            print()
    
    def is_valid_move(self, x, y):
        return 0 <= x < self.WIDTH and 0 <= y < self.HEIGHT and self.map[y][x] != self.WALL
    
    def move_player(self, dx, dy):
        newX = self.playerX + dx
        newY = self.playerY + dy
        
        if self.is_valid_move(newX, newY):
            self.map[self.playerY][self.playerX] = self.FLOOR
            self.playerX = newX
            self.playerY = newY
            self.map[self.playerY][self.playerX] = self.PLAYER
            self.update_fov()
    
    def getch(self):
        if os.name == 'nt':
            import msvcrt
            return msvcrt.getch().decode('utf-8')
        else:
            import sys
            import tty
            import termios
            fd = sys.stdin.fileno()
            old_settings = termios.tcgetattr(fd)
            try:
                tty.setraw(sys.stdin.fileno())
                ch = sys.stdin.read(1)
            finally:
                termios.tcsetattr(fd, termios.TCSADRAIN, old_settings)
            return ch
    
    def process_input(self):
        try:
            input_char = self.getch().lower()
            
            if input_char == 'w':
                self.move_player(0, -1)
            elif input_char == 's':
                self.move_player(0, 1)
            elif input_char == 'a':
                self.move_player(-1, 0)
            elif input_char == 'd':
                self.move_player(1, 0)
            elif input_char == 'q':
                self.gameRunning = False
                return True
        except:
            pass
        return False
    
    def run(self):
        while self.gameRunning:
            self.render()
            if self.process_input():
                break
            time.sleep(0.05)
        print("Спасибо за игру!")

if __name__ == "__main__":
    game = Roguelike()
    game.run()
